File: src/agent/catalog.ts
import { DEFAULT_BACKEND_ID, type PermissionMode } from './types';

/**
 * 后端 catalog —— 后端的「元数据 + 管理面驱动」单一真源（依赖类型/体积/检测/安装/
 * 分组/文案）。运行时工厂表（REGISTRY，见 index.ts）只负责构造能跑的实例；catalog
 * 负责「Web / DM 怎么展示、怎么装、怎么探」。两者职责正交、不合并、靠 id 配对
 * （单测强制 catalog.id 集合 == REGISTRY id 集合，新后端漏一处即红——防回归）。
 *
 * 「加一个新 agent CLI 后端」= 实现 AgentBackend + REGISTRY 加工厂 + 这里加一条
 * catalog（设计 §3.5）。Web 后端页 / DM picker / doctor / 按需下载按钮全自动出现。
 */

/** 底层 agent 家族（picker 按此分组；当前仅 Codex 组；string 为未来 agent 预留）。 */
export type AgentFamily = 'codex' | 'claude' | (string & {});

/** 后端进程的接入方式（仅描述/分组用，不参与运行路由）。 */
export type BackendAccess = 'app-server' | 'sdk' | 'acp';

/**
 * 依赖类型 —— 决定「装哪 / 怎么检测 / 能不能一键按需装」。
 *   'external-cli'  外部 CLI（codex / 未来 gemini-cli），bridge 不负责装，doctor 探 PATH。
 *   'npm-ondemand'  npm 包，按需装到用户私装目录（库类 / bin 类两形态）。
 *                   **唯一可一键下载的类型。** 库类（无 binName）走 import + require.resolve；
 *                   bin 类（有 binName）被 spawn、走 node_modules/.bin 路径（见 backend-loader）。
 *                   当前内置后端均非此类（codex 是 external-cli），保留以备将来挂新后端。
 *   'npm-external'  外部 npm 包，用户自管（当前内置后端无此类，保留给未来不便按需装的包）。
 */
export type DepKind = 'external-cli' | 'npm-ondemand' | 'npm-external';

export interface BackendDep {
  kind: DepKind;
  /** npm 包名（npm-ondemand / npm-external 时）。 */
  pkg?: string;
  /**
   * 该包作为「被 spawn 的可执行文件」消费时的 bin 名（npm 装包生成 node_modules/.bin/<binName>）。
   * 有此字段 ⇒ bin 类后端：已装判定/命令解析走 .bin 路径而非 require.resolve
   *   （bin-only 包通常无 main 入口，resolve 必失败）。
   * 无此字段 ⇒ 库类后端：走 import() + require.resolve。
   */
  binName?: string;
  /** pin 版本（npm-ondemand，避免漂移）；undefined ⇒ latest。 */
  version?: string;
  /** 体积提示 MB（Web 下载确认用，给用户预期）。 */
  approxSizeMB?: number;
  /** 检测/装法的一句话提示（external-cli 探不到、npm-external 未装时给用户）。 */
  detectHint: string;
  /** 安装命令的一句话说明（installable=false 的手动装法 / installable=true 的旁注）。 */
  installCmd?: string;
}

export interface BackendCatalogEntry {
  /** 与 REGISTRY id 一一对应（单测强制配对）。 */
  id: string;
  agentFamily: AgentFamily;
  displayName: string;
  access: BackendAccess;
  dep: BackendDep;
  /** 本后端支持的权限档（= backend.supportedModes）；undefined ⇒ 全档（codex）。 */
  supportedModes?: readonly PermissionMode[];
  /** picker 副标题（一句话接入说明）。 */
  blurb?: string;
  /**
   * 「用户不可见」闸（通用机制：把尚未就绪的后端先对用户全隐藏；当前无隐藏后端，
   * 保留以备将来挂新后端）。
   * true ⇒ 不进任何**用户可见面**：Feishu 新建/绑定卡的后端选择器、Web 后端页、
   *        宿主机体检页、智能默认后端的候选集。但**仍保留在 catalog/REGISTRY**
   *        （catalogBackendIds 仍含它 → 与 REGISTRY 的配对单测不破；代码在仓、
   *        只是无入口可达）。要点亮：把对应条目的 `hidden: true` 删掉即可，零返工。
   * 设计：曝光过滤集中在 {@link projectCreatableBackends} / pickDefaultBackend(detect.ts) /
   *      listBackendCatalog(admin/service.ts) / doctorBackends(admin/host.ts) 四处。
   */
  hidden?: boolean;
}

/**
 * 一条 catalog 是否可一键按需下载：npm-ondemand 才可（external-cli / npm-external
 * 要用户手动装）。Web 据此决定「未安装」时出「下载」按钮还是「手动装法提示」。
 */
export function isInstallable(entry: BackendCatalogEntry): boolean {
  return entry.dep.kind === 'npm-ondemand';
}

export const BACKEND_CATALOG: readonly BackendCatalogEntry[] = [
  {
    id: 'codex-appserver',
    agentFamily: 'codex',
    displayName: 'Codex',
    access: 'app-server',
    dep: {
      kind: 'external-cli',
      pkg: 'codex',
      detectHint: '未找到 codex CLI（设 CODEX_BIN、装 Codex.app，或 npm i -g @openai/codex）',
      installCmd: 'npm i -g @openai/codex（或装 Codex.app / 设 CODEX_BIN），然后 codex login',
    },
    // supportedModes undefined ⇒ 全档（qa/write/full）。
    blurb: '能力最全（goal/steer/compact/resume + 真沙箱只读档）',
  },
];

/**
 * 用户可见的后端 catalog（滤掉 {@link BackendCatalogEntry.hidden}）——Web 后端页 /
 * 宿主机体检页等「用户可见面」的列举从此取，而非裸 BACKEND_CATALOG。catalogBackendIds /
 * REGISTRY 配对仍走全量（hidden 只挡曝光、不退注册）。
 */
export function visibleCatalog(): BackendCatalogEntry[] {
  return BACKEND_CATALOG.filter((e) => !e.hidden);
}

/** 按 id 取一条 catalog（未注册返回 undefined）。 */
export function catalogById(id: string): BackendCatalogEntry | undefined {
  return BACKEND_CATALOG.find((e) => e.id === id);
}

/** 按 agent 家族取 catalog（picker 分组用）。 */
export function catalogByFamily(family: AgentFamily): BackendCatalogEntry[] {
  return BACKEND_CATALOG.filter((e) => e.agentFamily === family);
}

/** catalog 声明的全部后端 id（index.ts 的 backendIds 从此派生 → catalog 是单一注册入口）。 */
export function catalogBackendIds(): string[] {
  return BACKEND_CATALOG.map((e) => e.id);
}

/**
 * 新建项目时「可选后端」（飞书新建/绑定卡的后端下拉数据源）。规则（产品定）：
 *   ① codex（DEFAULT_BACKEND_ID）始终可选 —— 它是 external-cli 基线（全局 codex / Codex.app），
 *      isBackendEntryInstalled 对 external-cli 恒 false，但作为默认后端必须始终能选；
 *   ② 其余后端「已下载」才列（isInstalled 注入，本模块不碰文件系统、便于单测）；
 *   ③ 再按项目权限档过滤：后端 supportedModes 不含该档则剔除（仅支持部分档的后端，
 *      在不支持的档下自然不出现）。
 * 卡片里下不了后端 → 未下载的直接不显示，引导去 Web「后端 Agent」页下载。
 */
export function projectCreatableBackends(
  mode: PermissionMode,
  isInstalled: (entry: BackendCatalogEntry) => boolean,
): BackendCatalogEntry[] {
  return BACKEND_CATALOG.filter((e) => {
    if (e.hidden) return false; // 用户不可见闸：隐藏后端不进 picker
    const installed = e.id === DEFAULT_BACKEND_ID || isInstalled(e);
    if (!installed) return false;
    if (e.supportedModes && !e.supportedModes.includes(mode)) return false;
    return true;
  });
}

File: internal/agent/catalog.go
package agent

// catalog.go —— 后端元数据单一真源（对齐 TS agent/catalog）。
//
// 与运行时工厂表（REGISTRY）职责正交：catalog 管「Web/DM 怎么展示、怎么装、怎么探」，
// REGISTRY 管「怎么构造能跑的实例」。靠 id 配对（Phase 1 codex 实现，claude 二期）。

// AgentFamily 底层 agent 家族（picker 分组用）。
type AgentFamily string

const (
	FamilyCodex  AgentFamily = "codex"
	FamilyClaude AgentFamily = "claude"
)

// BackendAccess 接入方式（仅描述用）。
type BackendAccess string

const (
	AccessAppServer BackendAccess = "app-server"
	AccessSDK       BackendAccess = "sdk"
	AccessACP       BackendAccess = "acp"
)

// DepKind 依赖类型。
type DepKind string

const (
	DepExternalCLI DepKind = "external-cli"
	DepNpmOnDemand DepKind = "npm-ondemand"
	DepNpmExternal DepKind = "npm-external"
)

// BackendDep 后端依赖描述。
type BackendDep struct {
	Kind         DepKind
	Pkg          string
	BinName      string
	Version      string
	ApproxSizeMB int
	DetectHint   string
	InstallCmd   string
}

// BackendCatalogEntry catalog 条目。
type BackendCatalogEntry struct {
	ID             string
	AgentFamily    AgentFamily
	DisplayName    string
	Access         BackendAccess
	Dep            BackendDep
	SupportedModes []PermissionMode // nil=全档
	Blurb          string
	Hidden         bool
}

// BackendCatalog 全部后端元数据。
var BackendCatalog = []BackendCatalogEntry{
	{
		ID:          DEFAULT_BACKEND_ID,
		AgentFamily: FamilyCodex,
		DisplayName: "Codex",
		Access:      AccessAppServer,
		Dep: BackendDep{
			Kind:       DepExternalCLI,
			Pkg:        "codex",
			DetectHint: "未找到 codex CLI（设 CODEX_BIN、装 Codex.app，或 npm i -g @openai/codex）",
			InstallCmd: "npm i -g @openai/codex（或装 Codex.app / 设 CODEX_BIN），然后 codex login",
		},
		Blurb: "能力最全（goal/steer/compact/resume + 真沙箱只读档）",
	},
	{
		ID:          "claude-agent",
		AgentFamily: FamilyClaude,
		DisplayName: "Claude",
		Access:      AccessACP, // Go 侧经 `claude` CLI 的 stream-json 子进程（等价 ACP）
		Dep: BackendDep{
			Kind:       DepExternalCLI,
			Pkg:        "claude",
			BinName:    "claude",
			DetectHint: "未找到 claude CLI（安装 Claude Code：`npm i -g @anthropic-ai/claude-code`，或下载 Claude Code）",
			InstallCmd: "npm i -g @anthropic-ai/claude-code，然后 claude 登录",
		},
		SupportedModes: []PermissionMode{PermissionQA, PermissionWrite, PermissionFull},
		Blurb:          "Claude Code（复用本机登录；qa/write 走 Claude 沙箱，能力较 Codex 精简）",
	},
}

// IsInstallable 是否可一键按需下载（仅 npm-ondemand）。
func IsInstallable(e BackendCatalogEntry) bool { return e.Dep.Kind == DepNpmOnDemand }

// VisibleCatalog 用户可见 catalog（滤掉 Hidden）。
func VisibleCatalog() []BackendCatalogEntry {
	return visibleFromList(BackendCatalog)
}

func visibleFromList(list []BackendCatalogEntry) []BackendCatalogEntry {
	out := make([]BackendCatalogEntry, 0, len(list))
	for _, e := range list {
		if !e.Hidden {
			out = append(out, e)
		}
	}
	return out
}

// CatalogByID 按 id 取条目。
func CatalogByID(id string) (BackendCatalogEntry, bool) {
	for _, e := range BackendCatalog {
		if e.ID == id {
			return e, true
		}
	}
	return BackendCatalogEntry{}, false
}

// CatalogByFamily 按家族取条目。
func CatalogByFamily(family AgentFamily) []BackendCatalogEntry {
	var out []BackendCatalogEntry
	for _, e := range BackendCatalog {
		if e.AgentFamily == family {
			out = append(out, e)
		}
	}
	return out
}

// CatalogBackendIDs 全部后端 id。
func CatalogBackendIDs() []string {
	out := make([]string, 0, len(BackendCatalog))
	for _, e := range BackendCatalog {
		out = append(out, e.ID)
	}
	return out
}

// ProjectCreatableBackends 新建项目可选后端。
//   - codex（DEFAULT_BACKEND_ID）恒可选（external-cli 基线）；
//   - 其余已装才列（isInstalled 注入，本函数不碰文件系统）；
//   - 按项目权限档过滤；
//   - Hidden 不进 picker。
func ProjectCreatableBackends(mode PermissionMode, isInstalled func(BackendCatalogEntry) bool) []BackendCatalogEntry {
	return projectCreatableFromList(BackendCatalog, mode, isInstalled)
}

func projectCreatableFromList(list []BackendCatalogEntry, mode PermissionMode, isInstalled func(BackendCatalogEntry) bool) []BackendCatalogEntry {
	var out []BackendCatalogEntry
	for _, e := range list {
		if e.Hidden {
			continue
		}
		installed := e.ID == DEFAULT_BACKEND_ID || isInstalled(e)
		if !installed {
			continue
		}
		if len(e.SupportedModes) > 0 && !modeIn(e.SupportedModes, mode) {
			continue
		}
		out = append(out, e)
	}
	return out
}

func modeIn(modes []PermissionMode, m PermissionMode) bool {
	for _, x := range modes {
		if x == m {
			return true
		}
	}
	return false
}
